### PR TITLE
Convert to TypeScript, `open` can accept optional override parameters

### PR DESCRIPTION
**Pull request by aren55555**
### What

- Major version bump `3.0.0`
- **Task**: Converted this repo to TypeScript
- **NEW**: Exposed options on the `open` method for overriding the `products` (aka scopes) prior to opening Finch Connect. Enables more dynamic permissions.
- **Task**: Remove babel, cleaned up Rollup configurations

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    // "noPropertyAccessFromIndexSignature": true,       /* Enforces using indexed accessors for keys declared using an indexed type. */
+    // "allowUnusedLabels": true,                        /* Disable error reporting for unused labels. */
+    // "allowUnreachableCode": true,                     /* Disable error reporting for unreachable code. */
+
+    /* Completeness */
+    // "skipDefaultLibCheck": true,                      /* Skip type checking .d.ts files that are included with TypeScript. */
+    "skipLibCheck": true,                                 /* Skip type checking all .d.ts files. */
+  },
+  "include": ["src/**/*"],
+  "exclude": ["node_modules", "example/**/*"]
```
Comment (andrewdhazlett) [inline tsconfig.json:104]: shouldn't need `exclude` if you have `include`

File: tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    // "composite": true,                                /* Enable constraints that allow a TypeScript project to be used with project references. */
+    // "tsBuildInfoFile": "./.tsbuildinfo",              /* Specify the path to .tsbuildinfo incremental compilation file. */
+    // "disableSourceOfProjectReferenceRedirect": true,  /* Disable preferring source files instead of declaration files when referencing composite projects. */
+    // "disableSolutionSearching": true,                 /* Opt a project out of multi-project reference checking when editing. */
+    // "disableReferencedProjectLoad": true,             /* Reduce the number of projects loaded automatically by TypeScript. */
+
+    /* Language and Environment */
+    "target": "es2016",                                  /* Set the JavaScript language version for emitted JavaScript and include compatible library declarations. */
+    // "lib": [],                                        /* Specify a set of bundled library declaration files that describe the target runtime environment. */
+    "jsx": "preserve",                                /* Specify what JSX code is generated. */
```
Comment (andrewdhazlett) [inline tsconfig.json:16]: don't think we need this anymore

File: src/index.ts [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+  onClose: () => void;
+  payrollProvider: string | null;
+  products: string[];
+  sandbox: boolean;
+  zIndex: number;
+};
+
+type OpenFn = (overrides?: Partial<Pick<ConnectOptions, 'products'>>) => void;
+
+// TODO(@aren55555): from Finch Connect figure out the structure of this message
```
[the remark this answers is not in the document]
Comment (aren55555) [reply src/index.ts:26]: Yes -- I still have to reconcile these types and ensure they are aligned. It's possible some of the fields may be optional IIRC.

File: src/index.ts [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+export const useFinchConnect = (options: Partial<ConnectOptions>): { open: OpenFn } => {
+  if (!options.clientId) throw new Error('must specify clientId in options for useFinchConnect');
+
+  const combinedOptions: ConnectOptions = {
+    clientId: '',
+    ...DEFAULT_OPTIONS,
+    ...options,
+
+  }
```
Comment (aren55555) [inline src/index.ts:65]: There's really a use case for zod here.

File: src/index.ts [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+  onSuccess: (e: SuccessEvent) => void;
+  onError: (e: ErrorEvent) => void;
+  onClose: () => void;
+  payrollProvider: string | null;
+  products: string[];
+  sandbox: boolean;
+  zIndex: number;
+};
+
+type OpenFn = (overrides?: Partial<Pick<ConnectOptions, 'products'>>) => void;
```
[the remark this answers is not in the document]
Comment (aneeshvaidya) [reply src/index.ts:24]: nice!

File: src/index.ts [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+      iframe.style.backgroundColor = 'none transparent';
+      iframe.style.border = 'none';
+      document.body.prepend(iframe);
+      document.body.style.overflow = 'hidden';
+    }
+  };
+
+  const close = () => {
+    const frameToRemove = document.getElementById(FINCH_CONNECT_IFRAME_ID);
+    if (frameToRemove && frameToRemove.parentNode) {
```
Comment (andrewdhazlett) [inline src/index.ts:117]: nit: use safe-access operator

File: src/index.ts [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+  onSuccess: noop,
+  onError: noop,
+  onClose: noop,
+  payrollProvider: null,
+  products: [],
+  sandbox: false,
+  zIndex: 999,
+};
+
+export const useFinchConnect = (options: Partial<ConnectOptions>): { open: OpenFn } => {
```
Comment (andrewdhazlett) [inline src/index.ts:83]: nit: mark specific properties of `ConnectionOptions` as optional in the type definition above

File: src/index.ts [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+  if (!options.clientId) throw new Error('must specify clientId in options for useFinchConnect');
+
+  const combinedOptions: ConnectOptions = {
+    clientId: '',
```
Comment (andrewdhazlett) [inline src/index.ts:87]: not necessary w/ the `throw` above

**Review by andrewdhazlett — APPROVED** [review]

**Review by aneeshvaidya — APPROVED** [review]

File: src/index.ts (hunk 8 above)
Comment (aren55555) [reply src/index.ts:87]: This is where the inferred type system doesn't work well. TSC wasn't smart enough to figure this branch out, I was forced into adding a `clientId` field such that I could suppress the linter warning which is entirely unjustified based on the prior line.

File: src/index.ts (hunk 7 above)
Comment (aren55555) [reply src/index.ts:83]: I'm thinking about fast following with a 3.1.x version which has input checking (say via zod/joi/yup) and a standard shape/interface which we own for all of this library's possible Errors.

Comment (andrewdhazlett) [reply src/index.ts:83]: it is nice that this currently doesn't have any dependencies, i think we can do some light-weight hand-rolled validation. in general libraries should be super conservative wrt dependencies

File: src/index.ts (hunk 8 above)
Comment (andrewdhazlett) [reply src/index.ts:87]: in that case removing the `Partial<...>` wrapper and marking some properties as optional is the way to go

**Pull request merged**